src/components/graph-sonification.tsx

```diff
@@ -0,0 +1,174 @@
+  const currentGraphIdRef = useRef<string | null>(null);
+
+  if (!pannerRef.current) {
+    pannerRef.current = new Tone.Panner(0).toDestination();
+  }
+
+  if (!synthRef.current) {
+    synthRef.current = new Tone.PolySynth().connect(pannerRef.current);
+  }
+
+  const removeROIAdornment = async (graphId: string) => {
+    await codapInterface.sendRequest({
+      action: "delete",
+      resource: `component[${graphId}].adornment`,
+      values: {
+        type: "Region of Interest"
+      }
+    });
+  };
+
+  useEffect(() => {
+    const fetchGraphData = async () => {
+      const res = await codapInterface.sendRequest({action: "get", resource: `component[${graphToSonify}]`}) as IResult;
+      const graphDetails = res.values;
+      if (!graphDetails) return;
+
+      if (currentGraphIdRef.current && currentGraphIdRef.current !== graphDetails.id) {
+        await removeROIAdornment(currentGraphIdRef.current);
+      }
+
+      currentGraphIdRef.current = graphDetails.id;
+
+      await codapInterface.sendRequest({
+        action: "create",
+        resource: `component[${graphDetails.id}].adornment`,
+        values: {
+          type: "Region of Interest",
+          primary: { "position": 0, "extent": 0.05 },
+          secondary: { "position": 0, "extent": "100%" }
+        }
+      });
+
+      const pitchAttr = graphDetails.yAttributeName;
+      const timeAttr = graphDetails.xAttributeName;
+
+      const allItemsRes = await getAllItems(graphDetails.dataContext);
+      const allItems = allItemsRes.values;
+
+      // Asign pitch and time values to the dataset items
+      // Do not include items that are missing values for either attributes
+      const validItems = pitchAttr && timeAttr
+        ? allItems.filter((item: CodapItem) => item.values[pitchAttr] !== "" && item.values[timeAttr] !== "")
+        : allItems.filter((item: CodapItem) => item.values[timeAttr] !== "");
+
+      const pitchValues = validItems.map((item: CodapItem) => item.values[pitchAttr]);
+      const timeValues = validItems.map((item: CodapItem) => item.values[timeAttr]);
+
+      const minTime = graphDetails.xLowerBound;
+      const maxTime = graphDetails.xUpperBound;
+      const timeRange = maxTime - minTime || 1;
+
+      const minPitch = graphDetails.yLowerBound;
+      const maxPitch = graphDetails.yUpperBound;
+      const pitchRange = maxPitch - minPitch || 1;
+
+      const timeFractions = timeValues.map((t: number) => (t - minTime) / timeRange);
+      const pitchFractions = pitchValues.map((p: number) => (p - minPitch) / pitchRange);
+
+      await Tone.start();
+      Tone.getTransport().cancel();
+      Tone.getTransport().stop();
+      Tone.getTransport().position = 0;
+
+      const totalDuration = 10;
+      const fractionGroups: Record<number, number[]> = {};
+      timeFractions.forEach((frac: number, i: number) => {
+        if (!fractionGroups[frac]) fractionGroups[frac] = [];
+        fractionGroups[frac].push(i);
+      });
```

Comment:
This is one example, but this whole `fetchGraphData` method does a number of processing steps on the dataset -- filtering, mapping to ranges, assigning pan values, etc.  It would be much easier to read & understand if these separate operations were each named and documented - some could be general utilities, like scaling a value based on a range.

Reply:
@bgoldowsky Good point. We have another PR in the wings that adds controls for the sonification playback. In that one we broke this function up into smaller pieces. So we'll leave it as is here since those changes will be coming soon.